pkg/karmadactl/unregister/unregister.go

```diff
@@ -0,0 +1,301 @@
+
+	// 1. build member cluster client
+	restConfig, err := apiclient.RestConfig(j.Context, j.KubeConfig)
+	if err != nil {
+		return fmt.Errorf("failed to read member cluster rest config: %w", err)
+	}
+	j.MemberClusterClient, err = apiclient.NewClientSet(restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to build member cluster clientset: %w", err)
+	}
+
+	// 2. build karmada control plane client (read kubeconfig from member cluster's secret which stores the karmada kubeconfig for karmada agent)
+	karmadaConfigSecret, err := j.MemberClusterClient.CoreV1().Secrets(j.Namespace).Get(context.TODO(), register.KarmadaConfigSecretName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get the secret which stores the karmada kubeconfig")
+	}
+	karmadaCfg, err := clientcmd.Load(karmadaConfigSecret.Data[register.KarmadaConfigSecretKey])
+	if err != nil {
+		return err
+	}
+	if specifiedKarmadaContext := j.getSpecifiedKarmadaContext(j.MemberClusterClient); specifiedKarmadaContext != "" {
+		karmadaCfg.CurrentContext = specifiedKarmadaContext
+	}
+	j.ControlPlaneClient, err = register.ToKarmadaClient(karmadaCfg)
+	if err != nil {
+		return err
+	}
+
+	return j.RunUnregisterCluster()
+}
+
+type obj struct{ Kind, Name, Namespace string }
+
+func (o *obj) ToString() string {
+	if o.Namespace == "" {
+		return fmt.Sprintf("%s/%s", o.Kind, o.Name)
+	}
+	return fmt.Sprintf("%s/%s/%s", o.Kind, o.Namespace, o.Name)
+}
+
+// RunUnregisterCluster unregister the pull mode cluster from karmada.
+func (j *CommandUnregisterOption) RunUnregisterCluster() error {
+	if j.DryRun {
+		return nil
+	}
+
+	// 1. delete the cluster object in host cluster that associates the unregistering cluster
+	if err := cmdutil.DeleteClusterObject(j.ControlPlaneClient, j.ClusterName, j.Wait, j.DryRun); err != nil {
+		klog.Errorf("Failed to delete cluster object. cluster name: %s, error: %v", j.ClusterName, err)
+		return err
+	}
+	klog.Infof("Successfully delete cluster object (%s) from control plane.", j.ClusterName)
+
+	// 2. delete resource created by karmada in member cluster
+	clusterResources := []obj{
+		// 2.1 delete rbac resource which should upload to control plane to serve the controller-manager and aggregated-apiserver
+		{Kind: "ClusterRole", Name: names.GenerateRoleName(names.GenerateServiceAccountName(j.ClusterName))},
+		{Kind: "ClusterRoleBinding", Name: names.GenerateRoleName(names.GenerateServiceAccountName(j.ClusterName))},
+		{Kind: "ServiceAccount", Namespace: j.ClusterNamespace, Name: names.GenerateServiceAccountName(j.ClusterName)},
+		{Kind: "Secret", Namespace: j.ClusterNamespace, Name: names.GenerateServiceAccountName(j.ClusterName)},
+		{Kind: "ServiceAccount", Namespace: j.ClusterNamespace, Name: names.GenerateServiceAccountName("impersonator")},
+		{Kind: "Secret", Namespace: j.ClusterNamespace, Name: names.GenerateServiceAccountName("impersonator")},
+		{Kind: "Namespace", Name: j.ClusterNamespace},
```

Comment:
Can this be extracted into a method? Like `func(j *CommandUnregisterOption) getClusterResources() []obj`

Reply:
thanks, but what are the benefits? Isn't this clearer now?

Reply:
Extracting this part of the code would help make the `RunUnregisterCluster` process clearer, preventing readers from being blocked by a large list of resources when reading the code. It would also make it easier to write unit tests. Of course, this is just my suggestion.